### ISSUE TITLE
The negative score is not very beginner friendly

### ISSUE THREAD

**Issue by PCManticore** [PyCQA/pylint#2399]
For the past months I've been under the impression that the negative score is not very friendly for beginners. It somewhat demotivating to run `pylint` over your project just to get a `-24/10.0` as a score, which would make `pylint` look like a super drastic tool, that's super nitpicky about everything.

I think we could do a better job here at welcoming everyone to using `pylint`. Instead of having a `0-10` score, we could check if other tools are doing something different and see what works best in general.

**Comment by sushobhit27** [comment]
agree, it is very unusual.

**Comment by PCManticore** [comment]
@Hornwitser That's interesting, thanks for posting the comment. It might be worth giving it a try if you have some time to send a PR. Regarding your second comment, you might be interested in #746 which unfortunately stalled, but would provide a solution to the sorting and inspecting the messages that pylint emits.

**Comment by DC4JG** [comment]
The /10 strongly implied a rating of 0 or 1 to 10 for me, until I got a negative rating and questioned the whole system, started searching and found this issue.

No idea if the scoring system is good or bad in the bigger picture, but I can tell you it ain't intuitive.

**Comment by jacobtylerwalls** [comment]
From FAQ:

> However, this option can be changed in the Pylint rc file. If having negative
values really bugs you, you can set the formula to be the maximum of 0 and the
above expression.

Any reason we wouldn't just do this now? I don't know why we would need to wait an unspecified amount of time for v.3.0, especially if it's so easy to get the old behavior back. Do folks really have CI pipelines where the passing scores is negative and a score 0.0 will unexpectedly pass?

One argument for doing this in v2.13.0 is that we merged #5521 -- now a run with a fatal error can score higher than a negative score.

**Comment by DanielNoord** [comment]
Although I'm generally for changing things sooner rather than later I think the `2.12` update showed me that plugins and other integrations of `pylint` use our internals in very particular ways.
I don't think we can guarantee that nobody is relying on the fact that scores can be negative. I would be hesitant to change this right now. Also because there is not that much benefit to us. Although unintuitive, the fact that scores can be negative is not really limiting us in working on or improving pylint.

**Comment by Pierre-Sassoulas** [comment]
Daniel is right about the fact that it might breaks things for some plugin mainteners / users. Sometime it's surprising.

![workflow](https://user-images.githubusercontent.com/5493666/147258919-6fa89c74-922c-4d7f-a172-ab8241553454.png)

But it's also making users feel bad (I remember I felt hurt with my first negative score 😄) and it's easy to change so we could change it. 

I was thinking about this lately and maybe a solution would be to add levels, i.e. if you have an error, you'll have between 0 and 2.5/10 (Or between 0 and 10 but you're at level 0 of code quality), if you do not have any errors you'll have between 2.5 and 10.0 (Or 0 and 10 at level 1 of quality). This could affect more than the score and could also be the philosophy behind a kind of guide to what should be done next for a particular user instead of just dumping thousands of messages on their legacy code base. We made a pretty lengthy comments chain about the next step in gamification of pylint with @cdce8p in the last MR before a release and rather stupidly I did not create an issue with the discussion. Now it's lost in a MR from a long time ago that I can't find again 😄 Let me know if you remember where / when we talked about that Marc :)

**Comment by cdce8p** [comment]
Didn't know it either, but a search for `gamification` seemed to work 🤷🏻‍♂️ https://github.com/PyCQA/pylint/pull/4538#issuecomment-853355683

**Comment by Hornwitser** [comment]
> I don't think we can guarantee that nobody is relying on the fact that scores can be negative.

The scoring algorithm is a config option (`--evalutation`), if someone depends on the exact value returned by pylint's score they can set it to use the old algorithm when it's changed.

**Comment by DanielNoord** [comment]
> The scoring algorithm is a config option (`--evalutation`), if someone depends on the exact value returned by pylint's score they can set it to use the old algorithm when it's changed.

I think I misinterpreted @jacobtylerwalls. Sorry about that! Changing the default score evaluation should be fine indeed, not allowing negative scores should not be done (for now). Thanks for pointing this out @Hornwitser 👍

**Comment by jacobtylerwalls** [comment]
Cool. Yeah. That's what I was thinking. #5521 already changed the default evaluation for fatal errors and so it would make for a more consistent release story to say that in tandem with that we also set the (default) score floor to 0. I can push a PR to keep discussion going, not to be pushy!

**Comment by Pierre-Sassoulas** [comment]
Opened https://github.com/PyCQA/pylint/issues/5604 for the follow-up.

**Comment by Hornwitser** [comment]
I don't think the implementation of taking max(0, score) to avoid negative scores is any better than the negative scoring. Now when you get a really bad score it'll be 0 and there's the possibility that you could still be given a score of 0 after doing lots of improvements. For example where the previous score would be -20, it's now 0. If you improve it to what would be -3 it's now still 0. For the algorithm I proposed what would be a -20 would become a 0.3, and what would be a -3 would become a 1.0

**Comment by jacobtylerwalls** [comment]
It's marginally better in the sense that a fatal error can produce a score of zero as of #5521, so introducing a fatal error could have raised your score from -20 to 0, which is bad. Your suggestion still has independent value, so if it were me and I wanted to pursue it I would most likely open a new issue or draft PR.

**Comment by Pierre-Sassoulas** [comment]
I agree with both of you, feel free to open an issue with a one liner version of your proposal @Hornwitser.
